### DISCUSSION TITLE
How to upgrade from PostgreSQL 12 to 15?

### DISCUSSION THREAD

**Discussion by pippo73** [ghostfolio/ghostfolio#2509] [Q&A] [answered]
**Description**

I'm trying to upgrade to PostGres 15 from the previous 12

**To Reproduce**

1.I've modified the yaml file 
` postgres:
    ports:
      - 5432:5432
    image: postgres:15
`
2. docker pull ghostfolio-development-postgres-1
3. docker compose -f ghostfolio.yaml up -d

**Expected behavior**

the pg db switched from12 to 15

**Logs**

```
$ docker compose -f ghostfolio.yaml up -d
[+] Running 2/2
 ✔ Container ghostfolio-development-redis-1     Healthy                                                                                                    0.0s
 ✘ Container ghostfolio-development-postgres-1  Error                                                                                                      0.0s
dependency failed to start: container ghostfolio-development-postgres-1 is unhealthy

```

```
$ docker logs ghostfolio-development-postgres-1
[...]
PostgreSQL Database directory appears to contain a database; Skipping initialization

2023-10-18 08:14:11.371 UTC [1] FATAL:  database files are incompatible with server
2023-10-18 08:14:11.371 UTC [1] DETAIL:  The data directory was initialized by PostgreSQL version 12, which is not compatible with this version 15.4 (Debian 15.4-2.pgdg120+1).

```

**Environment**

- Self-hosted
- Ghostfolio Version v2.12.0.
- OS linux

**Additional context**

I cannot understand how to upgrade the pg db format from 12 to 15

**Comment by flowcool** [comment, marked as answer]
Hello @pippo73 

What I have done:

Put down ghostfolio frontend.

Perform a back-up:
`docker exec -it [your DB container name aka postgres] pg_dump  -U [YOUR USER] -d ghostfolio-db > [PUT YOUR Home directory]/upgrade_backup.sql`
**/!\ Put a copy somewhere in case of anything
/!\ read everything before executing blindling what I have done. If you're not confortable; don't do it.**

if you followed exactly the official conf:
`docker exec -it postgres pg_dump -U user -d ghostfolio-db > [PUT YOUR Home directory]/upgrade_backup.sql`

You can perform a backup of your volume before anything too:
`docker volume create --name ghostfolio_postgres_backup && docker run --rm -it -v ghostfolio_postgres:/from -v ghostfolio_postgres_backup:/to alpine ash -c 'cd /from ; cp -av . /to' && docker volume rm ghostfolio_postgres`
Delete old volume `ghostfolio_postgres`
Create a new volume for v15 postgres (personally I'm not using volume, so I created another folder)

We now have to initialize the new DB but avoid Ghostfolio to perform operation. Start your postgres container only.
either:
Comment out all services in the docker-compose file except for the postgres service
`docker-compose up`

or 

start postgres via portainer/GUI

-> Now the new postgres is initialized but some adjustments needs to be done.

put down DB 
-> in the file: postgresql.conf (/var/lib/postgresql/data/)
line 97; add:
`password_encryption = md5`

-> in the file: pg_hba.conf (/var/lib/postgresql/data/)
Replace :
`host all all all scram-sha-256`
by
`host all all all md5`

Then, we'll import back data:

`cat [PUT YOUR Home directory]/upgrade_backup.sql | docker exec -i postgres psql -U user -d ghostfolio-db`

Now you imported you data back to postgres. (Congrats)

Uncomment the other services in docker-compose and launch the application
Enjoy

(@MatanRubin Thanks for comments )

**Reply by MatanRubin** [reply]
@flowcool, thank you for the answer. I think there are some inaccuracies and missing instructions in it. Here's what I had to do to get this to work.

First, the command for dumping the db to a backup is missing "pg_dump". This is the right command:

```sh
docker exec -it [postgres-container-name] pg_dump -U user -d ghostfolio-db > [path to backup file]
```

The full paths to the files you need to edit are:

```
/var/lib/postgresql/data/postgresql.conf
/var/lib/postgresql/data/pg_hba.conf
```

The most important missing part is that you need to import the data back _before_ running the application itself for the first time. The reason is that the application initializes the DB one startup. Once the DB is initialized, you won't be able to import the backup file because there will be many conflicts and errors.

To get this to work:

1. Deleted all ghostfolio containers and volumes (after backing up your data)
2. Comment out all services in the docker-compose file except for the postgres service
3. docker-compose up
4. Import the backup file using the command in the original post
5. Uncomment the other services in docker-compose and launch the application

Hopefully this helps future readers.
@flowcool , I think you should amend your original answer with these details so that we can have one good tutorial for this upgrade.

**Reply by MatanRubin** [reply]
@flowcool , another addition to your post might be the way to backup the original volume (in addition to dumping the data to an sql file).
This can be done using the following docker command:

```sh
docker volume create --name ghostfolio_postgres_backup && docker run --rm -it -v ghostfolio_postgres:/from -v ghostfolio_postgres_backup:/to alpine ash -c 'cd /from ; cp -av . /to' && docker volume rm ghostfolio_postgres
```

This command would move all the data from the volume `ghostfolio_postgres` to the volume `ghostfolio_postgres_backup`, and then delete the volume `ghostfolio_postgres` (so that it could be created again using postgres15).

I liked the idea of having a double backup for the data, because you can never know what will go wrong.
After the entire operation was successful, I deleted the volume `ghostfolio_postgres_backup`.

**Reply by flowcool** [reply]
Thanks ! I amended it; but a bit blindly as I never use volume....so I hope it will be fine ;)
